tsdb/ooo_head_read.go

```diff
@@ -37,26 +37,29 @@ var _ IndexReader = &OOOHeadIndexReader{}
 // decided to do this to avoid code duplication.
 // The only methods that change are the ones about getting Series and Postings.
 type OOOHeadIndexReader struct {
-	*headIndexReader // A reference to the headIndexReader so we can reuse as many interface implementation as possible.
+	*headIndexReader     // A reference to the headIndexReader so we can reuse as many interface implementation as possible.
+	lastForbiddenMmapRef chunks.ChunkDiskMapperRef
 }
 
-func NewOOOHeadIndexReader(head *Head, mint, maxt int64) *OOOHeadIndexReader {
+func NewOOOHeadIndexReader(head *Head, mint, maxt int64, lastForbiddenMmapRef chunks.ChunkDiskMapperRef) *OOOHeadIndexReader {
 	hr := &headIndexReader{
 		head: head,
 		mint: mint,
 		maxt: maxt,
 	}
-	return &OOOHeadIndexReader{hr}
+	return &OOOHeadIndexReader{hr, lastForbiddenMmapRef}
 }
 
 func (oh *OOOHeadIndexReader) Series(ref storage.SeriesRef, builder *labels.ScratchBuilder, chks *[]chunks.Meta) error {
-	return oh.series(ref, builder, chks, 0)
+	return oh.series(ref, builder, chks, oh.lastForbiddenMmapRef, 0)
 }
 
-// The passed lastMmapRef tells upto what max m-map chunk that we can consider.
-// If it is 0, it means all chunks need to be considered.
-// If it is non-0, then the oooHeadChunk must not be considered.
-func (oh *OOOHeadIndexReader) series(ref storage.SeriesRef, builder *labels.ScratchBuilder, chks *[]chunks.Meta, lastMmapRef chunks.ChunkDiskMapperRef) error {
+// lastForbiddenMmapRef gives the last mmap chunk that may be being garbage collected and so
```

Comment:
I'm wondering if this should be `minMmapRef`. In a way we are constraining the list of chunks that can be considered for the query and for the future reader min and max may sound more easier to understand. Then the rest of the comment provide the context explaining where do these constraints come from

Reply:
I originally played with the idea of calling it `minMmapRef`, but I thought that would be confusing: `minMmapRef` to me sounds like it is the first chunk that can be considered (ie. it's inclusive), but it's the chunk before the first chunk that can be considered (ie. it's exclusive).

Maybe `lastGarbageCollectedMmapRef` would be clearer?

Reply:
Oh you are right. 

> Maybe lastGarbageCollectedMmapRef would be clearer?

Yeah lets do it.

Reply:
Done, I've changed the name in 4a14a44.